core/modules/replica/DatabaseMySQL.cc

```diff
@@ -664,4 +691,110 @@ void Connection::_assertTransaction(bool inTransaction) const {
     }
 }
 
+
+ConnectionPool::Ptr ConnectionPool::create(ConnectionParams const& params,
+                                           size_t maxConnections) {
+    return Ptr(new ConnectionPool(params, maxConnections));
+}
+
+
+ConnectionPool::ConnectionPool(ConnectionParams const& params, size_t maxConnections)
+    :   _params(params),
+        _maxConnections(maxConnections) {
+}
+
+
+Connection::Ptr ConnectionPool::allocate() {
+
+    string const context = "ConnectionPool::" + string(__func__) + "  ";
+
+    LOGS(_log, LOG_LVL_DEBUG, context);
+
+    unique_lock<mutex> lock(_mtx);
+
+    // Open a new connection and return it right away if the limit hasn't
+    // been reached yet.
+    //
+    // TODO: the factory method called below may put a calling thread
+    // in the blocking state while the (database) service becomes available.
+    // This will prevent operations with the pool by other threads. Investigate
+    // a non-blocking algorithm.
+
+    if (_availableConnections.size() + _usedConnections.size() < _maxConnections) {
```

Comment:
Can you reuse a connection from `_availableConnections` if that is non-empty?

Reply:
Great suggestion! Thanks! I will implement it.